\documentclass[12pt]{article}

\title{Computational biology in the 21st century:\\Algorithms that scale\\Supplement}
\date{}

\begin{document}
        
\maketitle

\section{Gene expression}
        
While sequence is important, scientists are also interested in measuring gene expression, or the quantity of
a particular gene product (usually a protein or more recently RNA) present in the cell.
Gene expression data is useful for relating genotype to phenotype, and is 
important in studying many diseases.
Usually obtained through microarrays or RNA-seq technologies, expression data is
quantitative, as each gene from a sample is associated with a numeric expression
level, and high-dimensional, as many thousands of genes may be analyzed at a
time.
Resources such as NCBIâ€™s Gene Expression Omnibus (GEO), which pool 
together many diverse expression studies, now allow researchers to analyze 
thousands of samples generated by others to obtain meaningful biological 
insights. 
More importantly, many of these biological insights can only be gleaned when 
viewed in the context of hundreds or thousands of gene expression samples~\cite{berger2013computational}.     

\section{Protein structure}   
        
Protein structure data, primarily determined through X-ray crystallography or
nuclear magnetic resonance (NMR) spectroscopy, describes the coordinates of 
each atom, in Angstroms, of a protein in three-dimensional space.
This information is important to understanding how a protein functions, because
proteins interact physically with other molecules and are geometrically constrained.
Protein structure data are commonly stored in the Protein Data Bank (PDB); 
each PDB entry has an associated resolution indicating the accuracy of the
structural description.
Thus a PDB entry at its most basic level contains a list of atoms, along with
which amino acid they belong to and their spatial coordinates as floating-point
numbers.        
        
Over the course of evolutionary time, structure is known to be more highly 
conserved than sequence~\cite{illergaard2009structure}, which is to say that it 
does not change as rapidly.
When the structure of a protein is known but its biological function or 
evolutionary relationships are not, researchers may search for structurally
similar proteins that are better studied~\cite{gibrat1996surprising}.
Classical tools for this involve performing pairwise structural alignments to
look for geometric similarity; DALI~\cite{holm1995dali} is still widely used,
along with other aligners such as FATCAT~\cite{ye2004fatcat} and 
Matt~\cite{menke2008matt}.
Due to the complexity of protein structures, these programs generally take 
significant amounts of time,
especially to align multiple structures.
For example, the DALI webserver can take as much as an hour to return results
for a single query.
FragBag~\cite{budowski2010fragbag} accelerates protein structure search by
approximating structural alignments
by instead comparing the `bag-of-words' from each structure.
Analogous to a term-frequency vector in information retrieval, this bag-of-words
indicates the abundance of particular, short structural motifs within a protein.



        
\bibliographystyle{abbrv}
\bibliography{main}

\end{document}